yellowbox/extras/logstash.py

```diff
@@ -1,30 +1,263 @@
+        delimiter = self.delimiter
+        encoding = self.encoding
+
+        def process_socket_data():
+            """Process & parse incoming socket data."""
+            data = sock.recv(1024)
+
+            # Socket closed
+            if not data:
+                self._selector.unregister(sock)
+                return
+
+            chunks = data.split(delimiter)
+
+            # Single partial chunk (no delimiter)
+            if len(chunks) == 1:
+                partial_chunks.append(chunks[0])
+                return
+
+            # Combine all partial chunks with first.
+            partial_chunks.append(chunks[0])
+            chunks[0] = b"".join(partial_chunks)
+            partial_chunks[:] = chunks.pop()
+
+            # Parse all chunks into records.
+            try:
+                for chunk in chunks:
+                    record_dict = json.loads(chunk.decode(encoding))
+                    self.records.append(record_dict)
+            except json.JSONDecodeError:
+                self._selector.unregister(sock)
+                sock.close()
+                # noinspection PyUnboundLocalVariable
+                _logger.exception("Failed decoding json, closing socket. "
+                                  "Data received: %s", chunk)
+                return
+
+        return process_socket_data
+
+    def _background_thread(self):
+        """Background thread processing incoming connections and data"""
+        while True:
+            events = self._selector.select(timeout=5)  # For signal handling
+            for key, mask in events:
+                # Handle closing request
+                if key.fileobj is self._rshutdown:
+                    assert self._rshutdown.recv(
+                        len(_CLOSE_SENTINEL)) == _CLOSE_SENTINEL
+                    return
+
+                # Handle new connection
+                if key.fileobj is self._root:
+                    new_socket, _ = self._root.accept()
+                    self._selector.register(new_socket, selectors.EVENT_READ,
+                                            self._create_data_callback(new_socket))
+                    continue
+
+                # Data received, run callback
+                try:
+                    key.data()
+                except Exception:
+                    _logger.exception("Unknown error occurred, closing connection.")
+                    self._selector.unregister(key.fileobj)
+                    # noinspection PyUnresolvedReferences
+                    key.fileobj.close()
+
+    def start(self, *, retry_spec: None = None) -> None:
+        """Start the service.
+
+        Args:
+            retry_spec: Ignored.
+        """
+        self._root.listen(_LISTEN_BACKLOG)
+        self._selector.register(self._root, selectors.EVENT_READ)
+        self._selector.register(self._rshutdown, selectors.EVENT_READ)
+        self._thread.start()
+        super(LogstashService, self).start(retry_spec=retry_spec)
```

Comment:
add a healthcheck mechanism, `start` should block until the listener thread is 100% receiving

Reply:
It is receiving what do you mean?

Reply:
that it accepts messages from loggers

Reply:
It does, why would it?

Reply:
because you might run into a race condition that you are proceeding with the test before the listener thread is actually in a listening state. `start` should avoid race conditions and block until it is certain the service can be used.

Reply:
You cant, it already listens beforehand.